### PR TITLE
Implement Swap via LIFI

### PR DESCRIPTION
**Pull request by justindg**
This branch most likely needs more refining - would love to get comments so I could further improve it.

Remaining key pieces:

- Translate gas fee into digestible information
- Send web3 transaction (Web3Transaction object is already built in SwapViewModel)

To test:
- On Wallet screen, click Ethereum token
- Click `More` (3 dot menu) -> Swap

### CODE REVIEW

**Comment by hboon** [comment]
Polyon is cheaper to test :P

**Review by JamesSmartCell — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/token_selector.xml [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/btn_select_token"
+            style="@style/Aw.Component.Button.Primary"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:drawableEnd="@drawable/ic_expand_more"
+            android:drawableTint="?colorSurface"
+            android:paddingHorizontal="@dimen/base_24"
+            android:text="Select Token" />
+
```
[the remark this answers is not in the document]
Comment (justindg) [reply app/src/main/res/layout/token_selector.xml:54]: Done

File: app/src/main/java/com/alphawallet/app/widget/ConfirmSwapDialog.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        return widget;
+    }
+
+    private SimpleSheetWidget buildFromWidget(Quote quote)
+    {
+        String srcAmt = BalanceUtils.getScaledValueFixed(new BigDecimal(quote.action.fromAmount), quote.action.fromToken.decimals, 4);
+        String srcTkn = quote.action.fromToken.symbol;
+        SimpleSheetWidget widget = new SimpleSheetWidget(getContext(), "From");
+        widget.setValue(srcAmt + " " + srcTkn);
+        return widget;
```
[the remark this answers is not in the document]
Comment (justindg) [reply app/src/main/java/com/alphawallet/app/widget/ConfirmSwapDialog.java:88]: Done

File: app/src/main/java/com/alphawallet/app/viewmodel/SwapViewModel.java [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        return result.contains("id")
+                && result.contains("action")
+                && result.contains("tool");
+    }
+
+    public String getBalance(String walletAddress, Connection.LToken token)
+    {
+        String address = token.address;
+        if (address.equalsIgnoreCase("0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee")) // TODO:
+        {
```
[the remark this answers is not in the document]
Comment (justindg) [reply app/src/main/java/com/alphawallet/app/viewmodel/SwapViewModel.java:273]: In LIFI, the native token's address is 0xeee or 0x000. In AlphaWallet's Token object the address is the wallet address, so this check is needed

File: app/src/main/java/com/alphawallet/app/widget/ConfirmSwapDialog.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    private SimpleSheetWidget buildFeeWidget(Quote quote)
+    {
+        SimpleSheetWidget widget = new SimpleSheetWidget(getContext(), "Fee");
+        widget.setValue("TODO"); // TODO: How to calculate this?
+        widget.setCaption("TODO"); // TODO
+        return widget;
+    }
+
```
[the remark this answers is not in the document]
Comment (justindg) [reply app/src/main/java/com/alphawallet/app/widget/ConfirmSwapDialog.java:72]: Gas fee is already shown in a separate widget. This is supposedly for other fees (AW commission?) but I don't think those are available yet

**Review by JamesSmartCell — APPROVED** [review]

File: app/src/main/java/com/alphawallet/app/widget/ConfirmSwapDialog.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (JamesSmartCell) [reply app/src/main/java/com/alphawallet/app/widget/ConfirmSwapDialog.java:72]: Ok gas estimate etc is already provided in the Quote return. I added it in already.

**Pull request merged**